File: app/src/main/java/com/hover/stax/countries/CountryDropdown.java
package com.hover.stax.countries;

import android.content.Context;
import android.util.AttributeSet;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.hover.stax.R;
import com.hover.stax.channels.Channel;
import com.hover.stax.utils.UIHelper;
import com.hover.stax.views.StaxDropdownLayout;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class CountryDropdown extends StaxDropdownLayout {

    CountryAdapter adapter;
    private CountryAdapter.SelectListener selectListener;

    public CountryDropdown(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public void updateChoices(List<Channel> channels, String currentCountry) {
        if (channels == null || channels.size() == 0) {
            setEmptyState();
            return;
        }
        adapter = new CountryAdapter(getCountryCodes(channels), getContext());
        autoCompleteTextView.setAdapter(adapter);
        autoCompleteTextView.setDropDownHeight(UIHelper.dpToPx(600));
        autoCompleteTextView.setOnItemClickListener((adapterView, view2, pos, id) -> onSelect((String) adapterView.getItemAtPosition(pos)));
        setDropdownValue(currentCountry);
        adapter.notifyDataSetChanged();
    }

    private String[] getCountryCodes(List<Channel> channelList) {
        List<String> countryCodes = new ArrayList<>();

        countryCodes.add(CountryAdapter.codeRepresentingAllCountries());
        for (Channel channel : channelList) {
            if (!countryCodes.contains(channel.countryAlpha2))
                countryCodes.add(channel.countryAlpha2);
        }
        Collections.sort(countryCodes);
        return countryCodes.toArray(new String[0]);
    }


    private void setEmptyState() {
        autoCompleteTextView.setDropDownHeight(0);
        setState(getContext().getString(R.string.channels_error_nodata), ERROR);
    }

    public void setListener(CountryAdapter.SelectListener sl) {
        selectListener = sl;
    }

    private void onSelect(String code) {
        setDropdownValue(code);
        if (selectListener != null) selectListener.countrySelect(code);
    }

    public void setDropdownValue(String countryCode) {
        if (countryCode == null)
            countryCode = CountryAdapter.codeRepresentingAllCountries();

        if (adapter != null)
            autoCompleteTextView.setText(adapter.getCountryString(countryCode));
    }
}

File: app/src/main/java/com/hover/stax/countries/CountryDropdown.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.countries

import android.content.Context
import android.util.AttributeSet
import com.hover.stax.R
import com.hover.stax.views.StaxDropdownLayout

class CountryDropdown(context: Context, attributeSet: AttributeSet) : StaxDropdownLayout(context, attributeSet) {

    private var countryAdapter: CountryAdapter? = null
    private var selectListener: CountryAdapter.SelectListener? = null

    fun updateChoices(countryList: List<String>, currentCountry: String?) {
        if (countryList.isEmpty()) {
            setEmptyState()
            return
        }

        countryAdapter = CountryAdapter(countryList.toTypedArray(), context)
        autoCompleteTextView.apply {
            setAdapter(countryAdapter)
            setOnItemClickListener { parent, _, position, _ -> onSelect(parent.getItemAtPosition(position) as String) }
        }

        setDropdownValue(currentCountry)
    }

    private fun setEmptyState() {
        setState(context.getString(R.string.channels_error_nodata), ERROR)
    }

    fun setListener(sl: CountryAdapter.SelectListener) {
        selectListener = sl
    }

    private fun onSelect(code: String) {
        setDropdownValue(code)
        selectListener?.countrySelect(code)
    }

    fun setDropdownValue(countryCode: String?) {
        autoCompleteTextView.setText(countryAdapter?.getCountryString(countryCode) ?: "")
    }
}